VideoLocker/src/main/java/org/edx/mobile/model/course/CourseStructureJsonHandler.java

```diff
@@ -0,0 +1,86 @@
+package org.edx.mobile.model.course;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.reflect.TypeToken;
+
+import org.edx.mobile.logger.Logger;
+
+import java.lang.reflect.Type;
+import java.util.Map;
+
+/**
+ * Created by hanning on 5/19/15.
+ */
+public class CourseStructureJsonHandler {
+    protected final Logger logger = new Logger(getClass().getName());
+
+     public CourseStructureV1Model processInput(String jsonInput)
+    {
+        BlockDataDeserializer dataDeserializer = new BlockDataDeserializer();
+        BlockListDeserializer blockListDeserializer = new BlockListDeserializer();
+        BlockTypeDeserializer blockTypeListDeserializer = new BlockTypeDeserializer();
+
+        //ideally we should use this approach. but it requires base class to
+        //define some properties.
+//        RuntimeTypeAdapterFactory<BlockData> rta = RuntimeTypeAdapterFactory.of(
+//            BlockData.class)
+//            .registerSubtype(VideoData.class);
+
+        Gson gson = new GsonBuilder()  //.registerTypeAdapterFactory(rta)
+            .registerTypeAdapter(BlockData.class, dataDeserializer)
+            .registerTypeAdapter(BlockType.class, blockTypeListDeserializer)
+            .registerTypeAdapter(BlockList.class, blockListDeserializer).create();
+
+        return gson.fromJson(jsonInput, CourseStructureV1Model.class);
+    }
+
+
+    class BlockDataDeserializer implements JsonDeserializer<BlockData>
+    {
+        @Override
+        public BlockData deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException
+        {
+            JsonObject jsonObject = json.getAsJsonObject();
+            //TODO - can not figure out a way to pass parent properties, for example, "type" field
```

Comment:
The internet suggests using thread local storage, but that doesn't really seem like a great solution either.